lib/spack/spack/test/error_messages.py

```diff
@@ -0,0 +1,554 @@
+# test_errmsg_requirements_2
+_pkgt4 = (
+    "t4",
+    """\
+class T4(Package):
+    version("2.1")
+    version("2.0")
+
+    variant("v1", default=True)
+
+    depends_on("t2")
+    depends_on("t2@:2.0", when="@:2.0")
+
+    depends_on("t3")
+    depends_on("t3~v1", when="@2.0")
+""",
+)
+
+
+_pkgt3 = (
+    "t3",
+    """\
+class T3(Package):
+    version("2.1")
+    version("2.0")
+
+    variant("v1", default=True)
+
+    requires("+v1", when="@2.1")
+
+    depends_on("t1")
+""",
+)
+
+
+_pkgt2 = (
+    "t2",
+    """\
+class T2(Package):
+    version("2.1")
+    version("2.0")
+
+    variant("v1", default=True)
+
+    # It used to be that swapping `require` for `conflicts` would
+    # improve the error message.
+    requires("~v1", when="@:2.0")
+    # conflicts("+v1", when="@:2.0")
+
+    depends_on("t1")
+""",
+)
+
+
+_pkgt1 = (
+    "t1",
+    """\
+class T1(Package):
+    version("2.1")
+    version("2.0")
+
+    variant("v1", default=True)
+""",
+)
+
+
+all_pkgs = [
+    _pkgx1,
+    _pkgx2,
+    _pkgx3,
+    _pkgx4,
+    _pkgy1,
+    _pkgy2,
+    _pkgy3,
+    _pkgy4,
+    _pkgz1,
+    _pkgz2,
+    _pkgz3,
+    _pkgw1,
+    _pkgw2,
+    _pkgw3,
+    _pkgw4,
+    _pkgt1,
+    _pkgt2,
+    _pkgt3,
+    _pkgt4,
+]
+
+
+def _add_import(pkg_def):
+    return (
+        """\
+from spack.package import *
+from spack.package import Package
+"""
+        + pkg_def
+    )
+
+
+all_pkgs = list((x, _add_import(y)) for (x, y) in all_pkgs)
+
+
+_repo_name_id = 0
+
+
+def create_test_repo(tmp_path, pkg_name_content_tuples):
+    global _repo_name_id
+
+    repo_name = f"testrepo{str(_repo_name_id)}"
+    repo_path = tmp_path / "spack_repo" / repo_name
+    os.makedirs(repo_path)
+    with open(repo_path / "__init__.py", "w", encoding="utf-8"):
+        pass
+    repo_yaml = os.path.join(repo_path, "repo.yaml")
+    with open(str(repo_yaml), "w", encoding="utf-8") as f:
+        f.write(
+            f"""\
+repo:
+  namespace: {repo_name}
+  api: v2.1
+"""
+        )
+
+    _repo_name_id += 1
+
+    packages_dir = repo_path / "packages"
+    os.mkdir(packages_dir)
+    with open(packages_dir / "__init__.py", "w", encoding="utf-8"):
+        pass
+    for pkg_name, pkg_str in pkg_name_content_tuples:
+        pkg_dir = packages_dir / pkg_name
+        os.mkdir(pkg_dir)
+        pkg_file = pkg_dir / "package.py"
+        with open(str(pkg_file), "w", encoding="utf-8") as f:
+            f.write(pkg_str)
+
+    repo_cache = spack.util.file_cache.FileCache(str(tmp_path / "cache"))
+    return spack.repo.Repo(str(repo_path), cache=repo_cache)
+
+
+@pytest.fixture
+def _create_test_repo(tmp_path, mutable_config):
+    yield create_test_repo(tmp_path, all_pkgs)
+
+
+@pytest.fixture
+def test_repo(_create_test_repo, monkeypatch, mock_stage):
+    with spack.repo.use_repositories(_create_test_repo) as mock_repo_path:
+        yield mock_repo_path
+
+
+@contextmanager
+def expect_failure_and_print(should_mention=None):
```

Comment:
None of the working tests actually use the `should_mention` argument here -- should we be adding that to the tests to make them more specific?